psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1785 @@
+}
+
+// Computes the LU decomposition of the coulomb metric and stores it in
+//   member vector CMPQ_LU_.
+//   pivots are storec in vector PERMUTE_
+void DirectDFJK::get_met() {
+    auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+    int info;
+
+    coul_met->form_fitting_metric();
+    SharedMatrix my_met;
+    my_met = coul_met->get_metric();
+    PERMUTE_.resize(naux_);
+    CMPQ_LU_.resize(naux_ * naux_);
+
+    CMPQ_inv_.resize(naux_ * naux_);
+
+//    double ZERO[8] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
+//    double* zerop = &ZERO[0];
+    std::vector<double> work_p(naux_);
+    //double* work_p =  &WORK_P[0];///.get();  // (double*)malloc(sizeof(double) * naux_);
+    // C_DCOPY( naux_, zerop, 0, work_p, 1);
+
+    memset(&work_p[0], 0, naux_ * 8);
+
+    double* cmpq_lup = &CMPQ_LU_.front();
+    double* cmpq_invp = &CMPQ_inv_.front();
+    double* metp = my_met->pointer()[0];
+    int* pert = &PERMUTE_.front();
+
+    info = C_DGETRF(naux_, naux_, metp, naux_, pert);
+    C_DCOPY(naux_ * naux_, metp, 1, cmpq_lup, 1);
+
+    C_DCOPY(naux_ * naux_, cmpq_lup, 1, cmpq_invp, 1);
+
+    C_DGETRI(naux_, cmpq_invp, naux_, pert, &work_p[0], naux_);
+
+//    if (work_p) {
+//        free(work_p);
+//    }
+}
+
+// called after sparsity_prep_pQq() 
+// is called in the memory estimate because the maximum permissable
+// block size is needed for an effective memory estimate, and determining
+// that is equivalent to estimating the total amount of memory
+// Determines the number of blocks and
+void DirectDFJK::prepare_p_blocks() {
+    Block_funcs_.resize(0);
+    Shell_starts_.resize(0);
+    Shell_stops_.resize(0);
+    k_disps_.resize(0);
+
+    size_t current_costs = 0;
+    // block max functions
+    size_t bmf = 0;
+
+    size_t block_size = 0;
+    size_t shell_funcs;
+    size_t biggest_shell = 0;
+
+    // one of two things. Either nelectron/2 + 1 or C_left_ao->ncol()
+    size_t charges_z;
+
+    /* Used because psi won't tell us the number of occupied orbitals until 
+     * the actual scf iterations, so we need to use this as a stand in.
+     * I'm not a huge fan either. */
+    if (C_left_ao_.size() == 0) {
+        // atomic charges float
+        double charges_f;
+
+        /*	for (size_t func_it = 0; func_it < schwarz_func_ints_.size(); func_it++) {
+                        sparse_fpf_ += schwarz_func_ints_[func_it];
+                }	*/
+
+        // We need to figure out the biggest shell in the primary basis set
+        // to allocate memory for AO construction
+        for (size_t i = 0; i < primary_->nshell(); i++) {
+            if (primary_->shell(i).nfunction() > biggest_shell) {
+                biggest_shell = primary_->shell(i).nfunction();
+            }
+        }
+
+        std::shared_ptr<Molecule> mol_ptr = primary_->molecule();
+
+        for (int atom_iter = 0; atom_iter < mol_ptr->natom(); atom_iter++) charges_f += mol_ptr->fZ(atom_iter);
```

Comment:
Do you have plans to account for non-neutral molecules? Please add a comment in the code if not.

Reply:
Unfortunately, this is the best I could do here. I believe i had had a comment there explaining what was going on, but I'll double check. The issue is that there's no way to know the actual dimensions of the orbital coefficient matrices every time this function is called, so I have to do what you see here to make sure that I'm asking psi to give the JK object enough memory to run the calculation. I'm using the molecule's charge as a way of guessing at the dimension of the coefficient matrix, and there's enough of a fudge factor here to make sure it doesn't cause any problems.

When the code gets called during the actual SCF iterations, the object does interrogate the coefficient matrices, so the non-neutral case is indeed handled.